core/gallery/cache.go

```diff
@@ -0,0 +1,231 @@
+	getVersion() (string, error)
+	save(index *pb.RpcGalleryDownloadIndexResponse, version string)
+}
+
+type storage struct {
+	versionPath, indexPath string
+}
+
+func (s *storage) getIndex() (*pb.RpcGalleryDownloadIndexResponse, error) {
+	rawData, err := os.ReadFile(s.indexPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read local gallery index: %w", err)
+	}
+
+	index := &pb.RpcGalleryDownloadIndexResponse{}
+	if err = proto.Unmarshal(rawData, index); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal local gallery index: %w", err)
+	}
+	return index, nil
+}
+
+func (s *storage) getVersion() (string, error) {
+	rawData, err := os.ReadFile(s.versionPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read local gallery index version: %w", err)
+	}
+	return string(rawData), nil
+}
+
+// TODO: Maybe we should save info to files async?
+func (s *storage) save(index *pb.RpcGalleryDownloadIndexResponse, version string) {
+	data, err := proto.Marshal(index)
+	if err != nil {
+		log.Error("failed to marshal local gallery index", zap.Error(err))
+		return
+	}
+
+	if err = os.WriteFile(s.indexPath, data, 0600); err != nil {
+		log.Error("failed to save local gallery index", zap.Error(err))
+		return
+	}
+
+	if err = os.WriteFile(s.versionPath, []byte(version), 0600); err != nil {
+		log.Error("failed to save local gallery version", zap.Error(err))
+	}
+}
+
+func (c *cache) downloadGalleryIndex(
+	timeoutInSeconds int, // timeout to wait for HTTP response
+	version string, // Etag of gallery index, that allows to fetch index faster
+	withManifestValidation bool, // a flag that indicates that every manifest should be validated
+) (response *pb.RpcGalleryDownloadIndexResponse, newVersion string, err error) {
+	raw, newVersion, err := getRawJson(c.indexURL, timeoutInSeconds, version)
+	if err != nil {
+		if errors.Is(err, ErrNotModified) {
+			return nil, version, err
+		}
+		return nil, "", fmt.Errorf("%w: %w", ErrDownloadIndex, err)
+	}
+
+	response = &pb.RpcGalleryDownloadIndexResponse{}
+	err = jsonpb.Unmarshal(bytes.NewReader(raw), response)
+	if err != nil {
+		return nil, "", fmt.Errorf("%w to get lists of categories and experiences from gallery index: %w", ErrUnmarshalJson, err)
+	}
+
+	if withManifestValidation {
+		for _, info := range response.Experiences {
+			if err = validateManifest(info.Schema, info); err != nil {
+				return nil, "", fmt.Errorf("manifest validation error: %w", err)
+			}
+		}
+	}
+
+	return response, newVersion, nil
+}
+
+func validateManifest(schema string, info *model.ManifestInfo) error {
+	if err := validateSchema(schema, info); err != nil {
+		return fmt.Errorf("manifest does not correspond scema: %w", err)
+	}
+
+	for _, urlToCheck := range append(info.Screenshots, info.DownloadLink) {
+		if !isInWhitelist(urlToCheck) {
+			return fmt.Errorf("URL '%s' provided in manifest is not in whitelist", urlToCheck)
+		}
+	}
+
+	info.Description = stripTags(info.Description)
```

Comment:
I think, it should be moved out of function, because it doesn't relate to validation, because according to name, this function is supposed to just validate manifest?🧐

Reply:
I have changed strip logic a lil bit. Now it accepts manifest and is not sticked to validation